### PR TITLE
Add Vue component for editing Heatmap and Griddata annotation elements

### LINKED ISSUE #135: Adjust colors of heatmap and griddata annotations
Expose controls to set the color range, range values, and normalizeRange of heatmap and griddata annotations.  For heatmap annotations, also allow changing the radius.  For griddata interpreted as a contour or choropleth, also allow changing the stepped flag, minColor, and maxColor.

For heatmaps, it would be good to expose the scale with plot option (this would need a boolean added to the schema).

It would be nice to display the min and max values in the dialog where this is done (or, ever better, some sort of histogram).

### PR DESCRIPTION
**Pull request by naglepuff**
Resolves #135 

This PR adds Vue to HistomicsUI and creates a component with Vue to edit attributes of heatmap and griddata annotation elements. To use this new component, simply click the gear icon next to the element in the draw widget when the annotation containing the element is selected.

### CODE REVIEW

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            });
+        },
+        notifySubmit() {
+            const propsToSave = {
+                rangeValues: this.colorRangeData.map((entry) => parseFloat(entry.value)),
+                colorRange: this.colorRangeData.map((entry) => entry.colorString),
+                normalizeRange: this.normalizeRange
+            };
+            if (this.type === 'heatmap') {
+                propsToSave['radius'] = parseInt(this.radius);
```
[the remark this answers is not in the document]
Comment (naglepuff) [reply histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:102]: Added radius validation in 855a464

**Comment by manthey** [comment]
Can we change the css so the color popup is on top rather than causing the dialog to need scrolling?
image

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (manthey) [reply histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:102]: `parseFloat` also is needed in the `notifySubmit` function.

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                                <tr>
+                                    <th>Value</th>
+                                    <th>Color</th>
+                                    <th></th>
+                                </tr>
+                            </thead>
+                            <tbody>
+                            <tr v-if="this.type === 'griddata'">
+                                <td>
+                                    Min. color
```
Comment (manthey) [inline histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:175]: Let's spell this and Max. Color out: "Minimum Color" and have a tool tip that says "The color of values below the lowest range value".

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+            return tinycolor(color).toString();
+        },
+        addColor(index) {
+            const newEntry = {
+                r: 0,
+                g: 0,
+                b: 0,
+                a: 0,
+                colorString: 'rgba(0, 0, 0, 0)',
+                value: 0
```
Comment (manthey) [inline histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:46]: When we add a color that is between two other colors, can we default it to a value halfway between the existing values?

Similarly, if it is a new first color, lets set the value to 0 and the color to whatever the lowest-rank color (excluding min/max( is but with a 0 opacity, and, if it is a new last color, the value should be the greater of 1 or the value of the highest-rank color record, and the color the same as the highest-rank color.  If the highest-rank opacity is 0, set it to 1.

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+            }
+            this.colorRangeData.splice(index + 1, 0, newEntry);
+            this.updateRangeDataKeys();
+        },
+        removeColor(index) {
```
Comment (manthey) [inline histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:51]: If there is only one color, don't allow its removal.

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                const invalidRangeValues = this.rangeValues.filter(
+                    (value) => value < 0 || value > 1).map(
+                    (value, index) => index + 1
+                );
+                if (invalidRangeValues.length) {
+                    this.validationErrors.push(
+                        'Invalid value(s) in row(s) ' + invalidRangeValues.join(', ') + '. When "Normalize Range" is checked, values must be between 0 and 1'
+                    );
+                }
+            }
```
Comment (manthey) [inline histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:85]: Values need to also be distinct.  As part of the submission, we should sort the list of range/colors by range value.

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (naglepuff) [reply histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue]: Added in 1ec048a

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue (hunk 2 above)
Comment (naglepuff) [reply histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue]: See 303d397

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue (hunk 3 above)
Comment (naglepuff) [reply histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:49]: Take a look at dd3938a for these changes.

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue (hunk 4 above)
Comment (naglepuff) [reply histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:81]: 8a1fc26

File: histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue (hunk 5 above)
Comment (naglepuff) [reply histomicsui/web_client/vue/components/EditHeatmapOrGridData.vue:130]: Added additional validation in 4be2b4f, added sorting in 42edea6

**Comment by naglepuff** [comment]
> Can we change the css so the color popup is on top rather than causing the dialog to need scrolling? image

See d1f108d

**Comment by manthey** [comment]
> > Can we change the css so the color popup is on top rather than causing the dialog to need scrolling? image
> 
> See [d1f108d](https://github.com/DigitalSlideArchive/HistomicsUI/commit/d1f108dc8758b86e49e05e32f4657f563a1bba23)

This still has an issue in that the top is cropped when editing the first color, but at least the control is useable.  Let's revisit that in a subsequent PR.

**Pull request merged**